Add a script for postprocessing author data

bockthom opened the pull request:
As some author disambiguation done by Codeface is just a heuristic, one might want to manually identify identical authors treated as different authors by Codeface. To unify them, specify a disambiguation list and update all author data in all .list files afterwards.

This fixes se-passau/coronet#178.

bockthom committed:
Add a script for postprocessing author data

As some author disambiguation done by Codeface is just a heuristic, one
might want to manually identfiy identical authors treated as different
authors by Codeface. To unify them, specify a disambiguation list and
update all author data in all .list files afterwards.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom commented:
Just one piece of information to add:

The manually created file `disambiguation-after-db.list` has the following structure:
One line represents two ids which should be unified to one person, that is, it is combined of two lines of the `authors.list`:
`id-to-keep; name-to-keep; email-to-keep;id-to-replace; name-to-replace; email-to-replace`
So, the first three columns specify the new person, the last three columns the person entry to replace by the new one.

If you want to unify three or more ids to one person, you need several lines - one for each entry to replace, having the same "to-keep" entry.

That's just an additional explanation as I forgot to place this information within the documentation. I will add this to the documentation later, but please review my PR first.

hechtlC commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -0,0 +1,217 @@
+
+            for person in disambiguation_data:
+                for author in authors:
+                    # replace author if nessary
```
necessary

hechtlC commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -0,0 +1,217 @@
+# Copyright 2020 by Thomas Bock <bockthom@cs.uni-saarland.de>
+# All Rights Reserved.
+"""
+This file is able to disambiguate authors after the extraction from the Codeface database was performed. A manually created disambiguation file is used to disambiguate the authors in all the extracted files of a project.
```
I don't know whether this is a GitHub problem or if this line is just really long. If that is the case, I think you should split the line for a more readable intro.

bockthom replied:
This line is really too long. I've just forgotten to add line breaks :wink: 
Thanks for spotting this @hechtlC !

I have already fixed this. In addition, I have added a more detailed description describing the format of the disambiguation file, including a small example.

hechtlC replied:
I just had a look at the new changes and didn't find any issues with it. So in my opinion this is ready for merging if @clhunsen approves.
Thanks again @bockthom!

hechtlC reviewed (approved):
First of all thank you very much for the enhancement @bockthom!
Aside from two small typo/formatting problems I don't see any issues here so I would be fine with merging.

bockthom committed:
Add more detailed documentation to author-postprocessing

Add description for format of the disambiguation file.

Fix typo and missing line break.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom committed:
Fix name in JIRA issues processing

When creating empty files at the beginning of the JIRA issue processing,
there was a file 'issue-jira.list' created which was never written to.
Instead, 'issues.list' was never cleared. This is fixed by clearing
'issues.list' instead of 'issue-jira.list'.

This fixes #33.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom committed:
JIRA issue processing: Don't run gephi export and archaical issues export

As the gephi export is not used and also not available for other extractions
or issue processing steps, don't run the gephi export anymore.
Same holds for the archaic issues export: Don't run it anymore as it is not
used and not compatible to other tools like coronet.

As these two exports are just commented out, still create empty files for them
at the beginning to avoid getting wrong results when someone wants to run these
commented exports but omitting clearing the corresponding output files.

Props to @ecklbarb for discussing whether the above mentioned exports
are still needed.

This is related to issue #33.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>
Signed-off-by: Barbara Eckl-Ganser <ecklbarb@fim.uni-passau.de>

clhunsen commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -0,0 +1,233 @@
+    - each line consists of six columns (each three for describing id, name, e-mail address)
+    - the entries of each line are taken from the global 'authors.list' file
+    - Example:
+        1234;Claus Hunsen;claus.hunsen@example.org;5678;claushunsen;claushunsen@hunsen.org;
```
Thanks for using my name in the example. :blush: Can you please adjust the example, which do *not* contain any possible top-level domains? For instance, you could use `@example.net` instead of `@hunsen.org`.

bockthom replied:
For this example, I was inspired by the author data of the coronet project. I used your name as you occurred twice in the coronet data and replaced the examples a little bit. (Another reason for using your name might be to especially motivate you to review this pull request :wink: )
Joking apart, I will replace the top-level domain as suggested by you. In addition, I will also swap surname and forename in the e-mail address in one of the examples to make the need for such a manual disambiguation more obvious (just replacing `org` by `net` and removing a `.` is a way too simple example).

clhunsen replied:
> For this example, I was inspired by the author data of the coronet project. I used your name as you occurred twice in the coronet data and replaced the examples a little bit. (Another reason for using your name might be to especially motivate you to review this pull request :wink: )

I thought so. :smile: 

> Joking apart, I will replace the top-level domain as suggested by you. In addition, I will also swap surname and forename in the e-mail address in one of the examples to make the need for such a manual disambiguation more obvious (just replacing `org` by `net` and removing a `.` is a way too simple example).

That is a really good idea. Thanks for the update.

clhunsen commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -0,0 +1,233 @@
+
+If more than two person identities should be mapped to each other, several lines are necessary in the disambiguation
+file. E.g., if persons A, B, C should be mapped to A, there has to be a line which replaces B by A (A,B) and a line
+which replaces C by A (A,C).
```
Overall, I really like your example! :+1:

clhunsen commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -0,0 +1,233 @@
+    - the first three columns of a line describe the person identity to keep (e.g., 1234)
+    - the last three columns of a line describe the person identity to replace (e.g., 5678)
+Result: Every occurrence of the second person identity will be replaced by the first person identity, in every .list
+file of the project (authors, commits, emails, issues, etc.)
```
Is there any backup of the existing files? If not, I would like to have this possibility (as a CLI switch, for example). For some projects, it really takes some time to re-extract the data from the database and having backup files may become handy.

bockthom replied:
> Is there any backup of the existing files?

Not yet. I just thought that as soon as we have run this new disambiguation step, nobody will be interested in the data from the extraction itself anymore. (And I won't apply this new step to the previous results' directories. So, it only will affect new data extractions.)

> If not, I would like to have this possibility (as a CLI switch, for example). For some projects, it really takes some time to re-extract the data from the database and having backup files may become handy.

A CLI switch sounds like a good idea (better than copying all results and wasting disk space per default). Most extractions are very fast, but you're right, there are some extractions which take some time.

> E.g., move all adjusted files to a timestamped backup folder while respecting their relative paths. thinking What do you think?

Hm. I'm not sure whether having timestamped backup folders is useful. This would create new backup folders whenever the extraction is run, resulting in many confusing backup folders. If you would like to have such, you can just run a backup script on the complete data. So, it is not necessary to integrate this into `codeface-extraction`.

When backuping the data by `codeface-extraction`, then I only have two ideas:
1. just copy and move the changed files, e.g., copy `authors.list` to `authors.bak.list` (or any other comprehensible name), for each of the affected files
2. copy the complete directory (e.g., `proximity` to `proximity.bak`) and only copy the `.list` files (to get rid of all the other cumbersome files created by `codeface`).

Both options, IMHO, fulfill all the requirements for such a backup. The second option also easily allows to integrate anonymization scripts in a similar way later on. So, the second option sounds more extensible and reusible to me (e.g., having a separate `proximity.anonymized` folder later -- but this are just dreams of the future, not to be discussed in this PR).

clhunsen replied:
> > E.g., move all adjusted files to a timestamped backup folder while respecting their relative paths. thinking What do you think?
> 
> Hm. I'm not sure whether having timestamped backup folders is useful. This would create new backup folders whenever the extraction is run, resulting in many confusing backup folders. If you would like to have such, you can just run a backup script on the complete data. So, it is not necessary to integrate this into `codeface-extraction`.
> 
> When backuping the data by `codeface-extraction`, then I only have two ideas:
> 
>     1. just copy and move the changed files, e.g., copy `authors.list` to `authors.bak.list` (or any other comprehensible name), for each of the affected files
> 
>     2. copy the complete directory (e.g., `proximity` to `proximity.bak`) and only copy the `.list` files (to get rid of all the other cumbersome files created by `codeface`).
> 
> 
> Both options, IMHO, fulfill all the requirements for such a backup. The second option also easily allows to integrate anonymization scripts in a similar way later on. So, the second option sounds more extensible and reusible to me (e.g., having a separate `proximity.anonymized` folder later -- but this are just dreams of the future, not to be discussed in this PR).

I thougth about the timestamped folders as repeated backups may overwrite existing backups and, thus, the original files. Apart from that, both ideas are good proposals, with the folder-based one being the "easier" one.

bockthom replied:
> I thougth about the timestamped folders as repeated backups may overwrite existing backups and, thus, the original files. 

That's true, good point! However, I still dislike the timestamped folders because of the additional maintenance effort to remove outdated folders and save disk space.

One option to mitigate the overwrite issue could be to perform the backup only if it is not present yet. That is, if the backup directory exists, don't perform the backup up if the backup flag is set. Nevertheless, this option also has its drawbacks: Even if the extraction from the database is run anew, the backup wouldn't be performed anew if the backup folder exists already. Either one has to remove the directory manually whenever a new backup should be created - or we automatically remove the backup folder whenever the extraction is run anew, but automatically removing backups is also a little bit scary - so I would prefer the manual effort of removing the backup folder (it could also be done automatically in external bash scripts but not within the extraction itself).

> Apart from that, both ideas are good proposals, with the folder-based one being the "easier" one.

Yeah, I also talked to @hechtlC and we agreed on the folder-based one. This also eases access to the backup: In coronet, for example, you only have to change one directory name to access the backuped data. So, I will implement that.

bockthom replied:
I have implemented the backup functionality in e9ab19a866edae7740f293103f0e2a40b87d0701, as suggested above.
@clhunsen: Could you please check if my implementation fulfils your requirements? Thanks!

clhunsen commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -0,0 +1,233 @@
+        if commits_list in filenames:
+            f = path.join(filepath, commits_list)
+            log.info("Postprocess %s ...", f)
+            authors = csv_writer.read_from_csv(f)
```
Please rename this variable to something like `commit_data` and adjust related loop-variables. Currently, it is quite difficult to distinguish between "persons" and "authors" and, correspondingly, interpret lines like `if person[4] == author[2] and person[5] == author[3]:`.

This holds analogously for the other files and data sources.

bockthom replied:
Good point. I was just too lazy to do that.
In the beginning, I just tested the script on the `authors.list` data and, therefore, I named this variable `author`. Afterwards, I applied this to all the other files via copy&paste, not changing the variable name because of possible later function extractions. Then I recognized that function extraction does not make sense due to too different columns accessed, but I was too lazy to rename the variables. But you are completely right, the variables should definitively be renamed :+1:

clhunsen reviewed (changes requested):
Thank you very much for this precise solution for the additional author disambiguation, @bockthom. :slightly_smiling_face::+1: 

I have two main remarks regarding the implementation:
- Is it possible to implement a backup functionality? E.g., move all adjusted files to a timestamped backup folder while respecting their relative paths. :thinking: What do you think?
- Some variable names (e.g., `authors`) are misleading and hinder code comprehensibility. I have annotated one occasiion explicitly.

bockthom committed:
Adjust variable names for better comprehensibility

The variable names in the author postprocessing script now refer to
the kind of data they represent.

Props to @clhunsen for pointing this out.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom committed:
Add backup option for author postprocessing

Add an additional CLI parameter to enable data backups before performing
the postprocessing. If data backups are enabled, all the  files of the
results directory are copied to another results directory having
suffix '_bak'. If this results directory does already exists, no backup
is performed even if backups are enabled.

Suggested-by: Claus Hunsen <hunsen@fim.uni-passau.de>
Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

clhunsen commented:
This looks very good. Thanks for your effort, @bockthom! :+1: 

@hechtlC, if you also approve the new changes, please merge this PR.

hechtlC commented on author_postprocessing/author_postprocessing.py:
```diff
@@ -51,16 +52,50 @@
 
+def perform_data_backup(results_path, results_path_backup):
+    """
+    Copy the existing .list files of a certain directory (also recursivly) to a separate backup folder.
```
"recursively"

hechtlC reviewed (approved):
This looks really good now. Thanks for this @bockthom! I only found one small typo and once this is fixed, I will merge the Pull Request.
